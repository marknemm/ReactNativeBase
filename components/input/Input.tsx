import { useFormControl, useFormErrorMessage, useValidationRules } from '@hooks/form-hooks';
import { useCallbacks } from '@hooks/state-hooks';
import { Input as RneInput, useTheme } from '@rneui/themed';
import { forwardRef, useEffect, useRef, useState } from 'react';
import { Controller } from 'react-hook-form';
import { useMaskedInputProps } from 'react-native-mask-input';
import type { InputFC, InputProps, InputRefType } from './Input.interfaces';
import { useStyles } from './Input.styles';

/**
 * Component for generic user text input.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link Input} component.
 */
const Input: InputFC = forwardRef((props, ref) => {
  const { label, name, onBlur, onChangeText } = props;

  // Derive entities related to an input controlled by react-hook-form
  const control = useFormControl(props);
  const rules = useValidationRules(props, label?.toString());
  const errorMessage = useFormErrorMessage(rules, props);

  return control
    ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange: onChangeForm, onBlur: onBlurForm, value } }) => (
          <InputControlled
            {...props}
            errorMessage={errorMessage}
            onBlur={(event) => {
              onBlurForm();
              onBlur?.(event);
            }}
            onChangeText={(text) => {
              onChangeForm(text);
              onChangeText?.(text);
            }}
            ref={ref}
            value={value}
          />
        )}
        rules={rules}
      />
    )
    : <InputControlled {...props} ref={ref} />;
});

/**
 * Controlled {@link Input} component.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link InputControlled} component.
 */
const InputControlled: InputFC = forwardRef(({
  containerStyle,
  inputContainerStyle,
  inputStyle,
  labelStyle,
  mask,
  maxLength,
  maxLengthLimitTyping,
  onChangeText,
  style,
  value,
  ...inputProps
}, ref) => {
  const styles = useStyles({ containerStyle, inputContainerStyle, inputStyle, labelStyle, style });
  const { theme } = useTheme();
  const [uiValue, setUiValue] = useState('');
  const inputRef = useRef<InputRefType>();
  ref ??= inputRef;

  const maskedInputProps = useMaskedInputProps({ mask, onChangeText, value });
  if (mask) {
    onChangeText = maskedInputProps.onChangeText;
    value = maskedInputProps.value;
  }

  const maxLengthNum = (typeof maxLength === 'number')
    ? maxLength
    : maxLength?.value ?? undefined;

  useEffect(() => {
    // Sync value prop with input (UI) value. Done manually to prevent change on each keystroke and prevent lag.
    if (value !== undefined && value !== uiValue) {
      const textInput = (ref as React.MutableRefObject<InputRefType>).current;
      textInput?.setNativeProps({ text: value });
    }
  }, [ref, value, uiValue]);

  return (
    <RneInput
      keyboardAppearance={theme.mode}
      placeholderTextColor={theme.colors.placeholder}
      {...inputProps}
      {...(mask ? maskedInputProps : undefined)}
      containerStyle={styles.container}
      inputContainerStyle={styles.inputContainer}
      inputStyle={styles.input}
      labelStyle={styles.label}
      maxLength={maxLengthLimitTyping ? maxLengthNum : undefined}
      onChangeText={useCallbacks(onChangeText, setUiValue)}
      ref={ref}
      style={styles.style}
      value={undefined} // Do not update value prop directly to prevent input lag (see useEffect above).
    />
  );
});

export type * from './Input.interfaces';
export default Input;
