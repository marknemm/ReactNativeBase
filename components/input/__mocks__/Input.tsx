import type { InputProps } from '@components/input/Input.interfaces';

const MockInput = jest.fn().mockImplementation((props: InputProps) => {
  const ActualComponent = jest.requireActual('@components/input/Input').default;
  return <ActualComponent {...props} />;
});

export default MockInput;
