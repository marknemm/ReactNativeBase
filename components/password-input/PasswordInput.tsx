import Input from '@components/input/Input';
import { PASSWORD_MIN_LENGTH_RULE } from '@constants/validation';
import type { IconNode } from '@rneui/base';
import { forwardRef, useMemo, useState } from 'react';
import type { PasswordInputFC, PasswordInputProps } from './PasswordInput.interfaces';

/**
 * An {@link Input} field for entering a password.
 *
 * @param props The component {@link PasswordInputProps}.
 * @param ref The component reference.
 * @returns The {@link PasswordInput} component.
 */
const PasswordInput: PasswordInputFC = forwardRef((props, ref) => {
  const { autoComplete, isVisibilityToggleEnabled = true, textContentType } = props;
  const [passwordVisible, setPasswordVisible] = useState(false);

  const derivedAutoComplete = autoComplete
    ?? (textContentType === 'password')
    ? 'current-password'
    : 'new-password';

  const derivedTextContentType = textContentType
    ?? (autoComplete === 'current-password')
    ? 'password'
    : 'newPassword';

  const passwordVisibleIcon = useMemo(
    () => (isVisibilityToggleEnabled
      ? {
        accessibilityLabel: 'Toggle password visibility',
        color: passwordVisible ? 'green' : 'gray',
        name: passwordVisible ? 'eye' : 'eye-off',
        type: 'feather',
        style: { padding: 7 },
        onPress: () => setPasswordVisible(!passwordVisible),
      } as IconNode
      : undefined
    ),
    [isVisibilityToggleEnabled, passwordVisible]
  );

  return (
    <Input
      autoCapitalize="none"
      autoCorrect={false}
      autoComplete={derivedAutoComplete}
      keyboardType={passwordVisible ? 'default' : 'visible-password'}
      minLength={PASSWORD_MIN_LENGTH_RULE}
      secureTextEntry={!passwordVisible}
      textContentType={derivedTextContentType}
      ref={ref}
      rightIcon={passwordVisibleIcon}
      {...props}
    />
  );
});

export type * from './PasswordInput.interfaces';
export default PasswordInput;
