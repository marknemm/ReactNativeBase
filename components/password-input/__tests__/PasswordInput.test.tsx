import Input from '@components/input/Input';
import PasswordInput from '@components/password-input/PasswordInput';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, screen, userEvent } from '@testing-library/react-native';

jest.mock('@components/input/Input');

describe('<PasswordInput />', () => {
  describe('password field type', () => {
    it('should derive default password field type', () => {
      render(
        <PasswordInput />,
        { wrapper: AppProvider }
      );

      expect(Input).toHaveBeenCalledWith(expect.objectContaining({
        autoComplete: 'new-password',
        textContentType: 'newPassword',
      }), expect.any(Object));
    });

    it('should derive password textContentType from autoComplete', () => {
      render(
        <PasswordInput autoComplete="current-password" />,
        { wrapper: AppProvider }
      );

      expect(Input).toHaveBeenCalledWith(expect.objectContaining({
        autoComplete: 'current-password',
        textContentType: 'password',
      }), expect.any(Object));
    });

    it('should derive password autoComplete from textContentType', () => {
      render(
        <PasswordInput textContentType="password" />,
        { wrapper: AppProvider }
      );

      expect(Input).toHaveBeenCalledWith(expect.objectContaining({
        autoComplete: 'current-password',
        textContentType: 'password',
      }), expect.any(Object));
    });
  });

  describe('password visibility', () => {
    it('should not show password by default', () => {
      render(
        <PasswordInput />,
        { wrapper: AppProvider }
      );

      expect(Input).toHaveBeenCalledWith(expect.objectContaining({
        rightIcon: expect.objectContaining({
          color: 'gray',
          name: 'eye-off',
        }),
        secureTextEntry: true,
      }), expect.any(Object));
    });

    it('should show password if visibility toggle is enabled', async () => {
      render(
        <PasswordInput />,
        { wrapper: AppProvider }
      );

      const visibilityToggle = screen.getByRole('button', { name: 'Toggle password visibility' });
      await userEvent.press(visibilityToggle);

      expect(Input).toHaveBeenCalledWith(expect.objectContaining({
        rightIcon: expect.objectContaining({
          color: 'green',
          name: 'eye',
        }),
        secureTextEntry: false,
      }), expect.any(Object));
    });

    it('should not show password visibility toggle if not enabled', () => {
      render(
        <PasswordInput isVisibilityToggleEnabled={false} />,
        { wrapper: AppProvider }
      );

      const visibilityToggle = screen.queryByRole('button', { name: 'Toggle password visibility' });
      expect(visibilityToggle).toBeFalsy();
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <PasswordInput />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
