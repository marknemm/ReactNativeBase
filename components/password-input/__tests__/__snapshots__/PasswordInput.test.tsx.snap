// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordInput /> snapshots renders correctly 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoComplete="new-password"
      autoCorrect={false}
      editable={true}
      keyboardAppearance="light"
      keyboardType="visible-password"
      minLength={
        {
          "message": "Password must be at least 6 characters",
          "value": 6,
        }
      }
      onChangeText={[Function]}
      placeholderTextColor="#86939e"
      secureTextEntry={true}
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      textContentType="newPassword"
      underlineColorAndroid="transparent"
    />
    <View
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "marginVertical": 4,
          "paddingRight": 4,
        }
      }
    >
      <View
        style={
          {
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityLabel="Toggle password visibility"
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "padding": 7,
            }
          }
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "gray",
                    "fontSize": 24,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              ï…¤
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;
